### ISSUE TITLE
DNP Master doesn't seem to recover from TCP errors

### ISSUE THREAD

**Issue by jadamcrain** [dnp3/opendnp3#62]
Issue is described here with log file from Verges

https://groups.google.com/forum/#!topic/automatak-dnp3/25j_dbERmqM

**Comment by cverges** [comment]
Latest test log from start of the system to lock-up:

```
2014-03-11 15:47:09.736 [scheduler-1] {} INFO  (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.master:-1) StackState: COMMS_DOWN
2014-03-11 15:47:09.736 [scheduler-1] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.Router:-1) Start()
2014-03-11 15:47:09.736 [scheduler-1] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.Router:-1) MonitorStateInit(CLOSED) -> MonitorStateOpening(OPENING) : Open: 0 Opening: 0 Reading: 0 Writing: 0 Closing: 0
2014-03-11 15:47:13.013 [Thread-9] {} INFO  (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1:-1) Connected to: 10.205.64.28:20000
2014-03-11 15:47:13.013 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.Router:-1) _OnLowerLayerUp
2014-03-11 15:47:13.013 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.Router:-1) MonitorStateOpening(OPENING) -> MonitorStateOpen(OPEN) : Open: 1 Opening: 0 Reading: 0 Writing: 0 Closing: 0
2014-03-11 15:47:13.014 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.transport:-1) State Change: TLS_Closed -> TLS_Ready
2014-03-11 15:47:13.014 [Thread-9] {} INFO  (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.master:-1) State Change: AMS_Closed -> AMS_Idle
2014-03-11 15:47:13.014 [Thread-9] {} INFO  (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.master:-1) State Change: AMS_Idle -> AMS_Waiting
2014-03-11 15:47:13.014 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.app.sol:-1) State changed from ACS_Idle to ACS_SendExpectResponse
2014-03-11 15:47:13.015 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.transport:-1) State Change: TLS_Ready -> TLS_Sending
2014-03-11 15:47:13.015 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.transport:-1) State Change: TLS_Sending -> TLS_Ready
2014-03-11 15:47:13.015 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.app.sol:-1) State changed from ACS_SendExpectResponse to ACS_WaitForFirstResponse
2014-03-11 15:47:13.695 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.app.sol:-1) State changed from ACS_WaitForFirstResponse to ACS_Idle
2014-03-11 15:47:13.697 [Thread-9] {} INFO  (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.master:-1) StackState: COMMS_UP
2014-03-11 15:47:13.762 [Thread-9] {} INFO  (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.master:-1) State Change: AMS_Waiting -> AMS_Idle
2014-03-11 15:47:13.763 [Thread-9] {} INFO  (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.master:-1) State Change: AMS_Idle -> AMS_Waiting
2014-03-11 15:47:13.763 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.app.sol:-1) State changed from ACS_Idle to ACS_SendExpectResponse
2014-03-11 15:47:13.763 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.transport:-1) State Change: TLS_Ready -> TLS_Sending
2014-03-11 15:47:13.763 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.transport:-1) State Change: TLS_Sending -> TLS_Ready
2014-03-11 15:47:13.763 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.app.sol:-1) State changed from ACS_SendExpectResponse to ACS_WaitForFirstResponse
2014-03-11 15:47:15.193 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.app.sol:-1) State changed from ACS_WaitForFirstResponse to ACS_Idle
2014-03-11 15:47:16.407 [Thread-9] {} INFO  (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.master:-1) State Change: AMS_Waiting -> AMS_Idle
2014-03-11 15:47:16.407 [Thread-9] {} INFO  (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.master:-1) State Change: AMS_Idle -> AMS_Waiting
2014-03-11 15:47:16.407 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.app.sol:-1) State changed from ACS_Idle to ACS_SendExpectResponse
2014-03-11 15:47:16.407 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.transport:-1) State Change: TLS_Ready -> TLS_Sending
2014-03-11 15:47:16.408 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.transport:-1) State Change: TLS_Sending -> TLS_Ready
2014-03-11 15:47:16.408 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.app.sol:-1) State changed from ACS_SendExpectResponse to ACS_WaitForFirstResponse
2014-03-11 15:47:16.913 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.app.sol:-1) State changed from ACS_WaitForFirstResponse to ACS_Idle
2014-03-11 15:47:16.913 [Thread-9] {} INFO  (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.master:-1) State Change: AMS_Waiting -> AMS_Idle
2014-03-11 15:47:16.913 [Thread-9] {} INFO  (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.master:-1) State Change: AMS_Idle -> AMS_Waiting
2014-03-11 15:47:16.913 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.app.sol:-1) State changed from ACS_Idle to ACS_SendExpectResponse
2014-03-11 15:47:16.914 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.transport:-1) State Change: TLS_Ready -> TLS_Sending
2014-03-11 15:47:16.914 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.transport:-1) State Change: TLS_Sending -> TLS_Ready
2014-03-11 15:47:16.914 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.app.sol:-1) State changed from ACS_SendExpectResponse to ACS_WaitForFirstResponse
2014-03-11 15:47:17.513 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.app.sol:-1) State changed from ACS_WaitForFirstResponse to ACS_Idle
2014-03-11 15:47:17.868 [Thread-9] {} INFO  (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.master:-1) State Change: AMS_Waiting -> AMS_Idle
2014-03-11 15:47:17.868 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.master:-1) TransmitVtoData: false size: 0
2014-03-11 16:47:17.868 [Thread-9] {} INFO  (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.master:-1) State Change: AMS_Idle -> AMS_Waiting
2014-03-11 16:47:17.869 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.app.sol:-1) State changed from ACS_Idle to ACS_SendExpectResponse
2014-03-11 16:47:17.869 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.transport:-1) State Change: TLS_Ready -> TLS_Sending
2014-03-11 16:47:17.870 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.transport:-1) State Change: TLS_Sending -> TLS_Ready
2014-03-11 16:47:17.870 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.app.sol:-1) State changed from ACS_SendExpectResponse to ACS_WaitForFirstResponse
2014-03-11 16:47:22.879 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.app.sol:-1) State changed from ACS_WaitForFirstResponse to ACS_Idle
2014-03-11 16:47:23.190 [Thread-9] {} INFO  (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.master:-1) State Change: AMS_Waiting -> AMS_Idle
2014-03-11 17:33:33.798 [Thread-9] {} INFO  (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.master:-1) State Change: AMS_Idle -> AMS_Waiting
2014-03-11 17:33:33.798 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.app.sol:-1) State changed from ACS_Idle to ACS_SendExpectResponse
2014-03-11 17:33:33.798 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.transport:-1) State Change: TLS_Ready -> TLS_Sending
2014-03-11 17:33:33.798 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.transport:-1) State Change: TLS_Sending -> TLS_Ready
2014-03-11 17:33:33.799 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.app.sol:-1) State changed from ACS_SendExpectResponse to ACS_WaitForFirstResponse
2014-03-11 17:33:53.799 [Thread-9] {} WARN  (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.app.sol:-1) Timeout while waiting for response
2014-03-11 17:33:53.799 [Thread-9] {} DEBUG (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.app.sol:-1) State changed from ACS_WaitForFirstResponse to ACS_Idle
2014-03-11 17:33:53.800 [Thread-9] {} INFO  (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.master:-1) StackState: COMMS_DOWN
2014-03-11 17:33:53.868 [Thread-9] {} INFO  (MasterStackManagerImpl.java:240) - DNP Log Message: (channel-tcp-client-endpoint-1.stack-master-endpoint-1.master:-1) State Change: AMS_Waiting -> AMS_Idle
```

Note that the physical layer received a TCP RST when attempting to set a latch output, which is what caused the timeout and subsequent COMMS_DOWN event.  After waiting 10 minutes, the stack does not appear to have recovered on its own.

**Comment by jadamcrain** [comment]
I believe that this was the bug in the master state machine if the layer close while sending a control. I believe this is fixed on 1.1.x and refactored on 2.0.x
